src/Microsoft.CodeQuality.Analyzers.Exp/Core/FlowAnalysis/Analysis/ParameterValidationAnalysis/ParameterValidationAnalysis.cs

```diff
@@ -0,0 +1,94 @@
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using Analyzer.Utilities.Extensions;
+using Microsoft.CodeAnalysis.Operations.ControlFlow;
+
+namespace Microsoft.CodeAnalysis.Operations.DataFlow.ParameterValidationAnalysis
+{
+    using ParameterValidationAnalysisData = IDictionary<AbstractLocation, ParameterValidationAbstractValue>;
+    using ParameterValidationAnalysisDomain = MapAbstractDomain<AbstractLocation, ParameterValidationAbstractValue>;
+
+    /// <summary>
+    /// Dataflow analysis to track <see cref="ParameterValidationAbstractValue"/> of <see cref="AbstractLocation"/>/<see cref="IOperation"/> instances.
+    /// </summary>
+    internal partial class ParameterValidationAnalysis : ForwardDataFlowAnalysis<ParameterValidationAnalysisData, ParameterValidationBlockAnalysisResult, ParameterValidationAbstractValue>
+    {
+        public static readonly ParameterValidationAnalysisDomain ParameterValidationAnalysisDomainInstance = new ParameterValidationAnalysisDomain(ParameterValidationAbstractValueDomain.Default);
+
+        private ParameterValidationAnalysis(ParameterValidationAnalysisDomain analysisDomain, ParameterValidationDataFlowOperationVisitor operationVisitor)
+            : base(analysisDomain, operationVisitor)
+        {
+        }
+
+        public static ImmutableDictionary<IParameterSymbol, SyntaxNode> GetOrComputeHazardousParameterUsages(
+            IOperation topmostBlock,
+            Compilation compilation,
+            ISymbol owningSymbol,
+            bool pessimisticAnalysis = true)
+        {
+            Debug.Assert(topmostBlock != null);
+
+            var wellKnownTypeProvider = WellKnownTypeProvider.GetOrCreate(compilation);
+            (DataFlowAnalysisResult<ParameterValidationBlockAnalysisResult, ParameterValidationAbstractValue> parameterValidationAnalysisResult, ImmutableDictionary<IParameterSymbol, SyntaxNode> hazardousParameterUsages) GetOrComputeLocationAnalysisResultForInvokedMethod(IBlockOperation methodTopmostBlock, IMethodSymbol method)
+            {
+                // getOrComputeLocationAnalysisResultOpt = null, so we do interprocedural analysis only one level down.
+                Debug.Assert(methodTopmostBlock != null);
+                return GetOrComputeResult(methodTopmostBlock, method, wellKnownTypeProvider, getOrComputeLocationAnalysisResultOpt: null, pessimisticAnalysis: pessimisticAnalysis);
+            };
+
+            var result = GetOrComputeResult(topmostBlock, owningSymbol, wellKnownTypeProvider, GetOrComputeLocationAnalysisResultForInvokedMethod, pessimisticAnalysis);
+            return result.hazardousParameterUsages;
+        }
+
+        private static (DataFlowAnalysisResult<ParameterValidationBlockAnalysisResult, ParameterValidationAbstractValue> parameterValidationAnalysisResult, ImmutableDictionary<IParameterSymbol, SyntaxNode> hazardousParameterUsages) GetOrComputeResult(
```

Comment:
>DataFlowAnalysisResult<ParameterValidationBlockAnalysisResult, ParameterValidationAbstractValue> parameterValidationAnalysisResult [](start = 24, length = 130)

As a stylistic thing, we might want to create some using aliases for these types. They're really long and everywhere.